hazelcast/src/main/java/com/hazelcast/query/impl/Indexes.java

```diff
@@ -204,6 +205,53 @@ public void createIndexesFromRecordedDefinitions(StoreAdapter partitionStoreAdap
         return indexes;
     }
 
+    /**
+     * Returns known and populated indexes for the given partition. If no
+     * indexes are known or populated, empty array is returned.
+     * <p>
+     * Must be called on partition threads only.
+     */
+    public InternalIndex[] getPopulatedIndexes(int partitionId) {
+        ThreadUtil.assertRunningOnPartitionThread();
```

Comment:
Needs to be rebased because of changes in `util` package.
Also, it can be called from non-partition threads. For instance:

`MapServiceContextImpl#reset`, `MapServiceContextImpl#shutdown` -> `MapServiceContextImpl#removeRecordStoresFromPartitionMatchingWith` -> `DefaultRecordStore#clearPartition` -> `DefaultRecordStore#clearOtherDataThanStorage` -> `DefaultRecordStore#clearIndexedData` -> `DefaultRecordStore#clearGlobalIndexes` -> `DefaultRecordStore#fullScanLocalDataToClear` -> `Indexes#getPopulatedIndexes`.

Can you check?

Reply:
I have tracked all possible (I hope) execution paths leading to `Indexes.getPopulatedIndexes`, the paths leading to partition thread execution are rooted at: `MapMigrationAwareService.commitMigration/rollbackMigration`, `MapSplitBrainHandlerService.onStoreCollection`, `MapPartitionDestroyOperation.run` and `EnterpriseMapPartitionClearOperation.run`.

The two problematic ones are `MapServiceContextImpl.shutdown`and `MapServiceContextImpl.reset`. The first one invokes `removeAllRecordStoresOfAllMaps` with `onShutdown = true`, so it's not leading to `Indexes.getPopulatedIndexes` invocation. The second one is invoked from `HotRestartIntegrationService.resetNode` and `ClusterMergeTask.resetServices`. To my understanding, `reset` can be invoked only when node is in a special turned off state, so we can assume there is only a single thread that can invoke `reset` (but I might be wrong on that). To make it even more fragile we can just remove `assertRunningOnPartitionThread`, WDYT? :)

HD is not affected since its `removeAllRecordStoresOfAllMaps` is overridden and it wraps all operations into `EnterpriseMapPartitionClearOperation` which runs on partition threads.

Reply:
You're right, I would also assume there can ever be only a single thread invoking `reset()`. Thing is, as you've noticed, on OS side it might not be the partition thread so this assertion might fail, e.g. when running with assertions enabled (even though nothing bad will happen). I would like to keep the assertion though, just don't know how to make it more resilient.